### PR TITLE
[CALCITE-4650] TopDownRuleDriver.DeriveTrait.derive produces wrong AssertionError when registering new nodes (Vladimir Ozerov)

### PR DESCRIPTION
**Pull request by devozerov**
See https://issues.apache.org/jira/browse/CALCITE-4650 for more details.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java [hunk 1]
```diff
@@ -952,7 +952,8 @@ private void derive() {
         List<RelNode> relList = rel.derive(inputTraits);
         for (RelNode relNode : relList) {
           if (!planner.isRegistered(relNode)) {
-            planner.register(relNode, rel);
+            RelSubset relSubset = planner.register(relNode, rel);
+            assert relSubset.set == VolcanoPlanner.canonize(planner.getSubsetNonNull(rel)).set;
           }
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:957]: Could you please add a test case?

File: core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java [hunk 2]
```diff
@@ -937,7 +937,7 @@ private void derive() {
                 subset.disableEnforcing();
               }
               RelSubset relSubset = planner.register(newRel, rel);
-              assert relSubset.set == planner.getSubsetNonNull(rel).set;
+              assert relSubset.set == VolcanoPlanner.canonize(planner.getSubsetNonNull(rel)).set;
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:940]: This looks really suspicious.
Should subset merge cleanup the references? Should `getSubsetNonNull` return the canonical subset?

Comment (hsyuan) [reply core/src/main/java/org/apache/calcite/plan/volcano/TopDownRuleDriver.java:940]: Agree with @vlsi.

**Comment by devozerov** [comment]
Hi @vlsi @hsyuan @chunweilei,
It's been a while since I opened this PR. I do not have a reproducer anymore. The problem appeared in the private codebase in one of the TPC-DS tests with a certain set of rules and a concrete `derive/passThrough` logic. Since then there were multiple changes and I no longer can reproduce the error. I cannot prove that it was a real problem with a planner as opposed to a problem with my own logic. Therefore, closing the PR and the ticket for now. Will re-open if needed.

**Pull request closed without merging**